pkg/scheduler/scheduler.go

```diff
@@ -114,3 +123,83 @@ type VPod interface {
 
 	GetResourceVersion() string
 }
+
+type ScaleCache struct {
+	lock                 sync.RWMutex //protects access to entries, entries itself is concurrency safe, so we only need to ensure that we correctly access the pointer
+	entries              *cache.Expiring
+	statefulSetClient    clientappsv1.StatefulSetInterface
+	statefulSetNamespace string
+}
+
+type scaleEntry struct {
+	specReplicas int32
+	statReplicas int32
+}
+
+func NewScaleCache(ctx context.Context, namespace string) *ScaleCache {
+	return &ScaleCache{
+		entries:              cache.NewExpiring(),
+		statefulSetClient:    kubeclient.Get(ctx).AppsV1().StatefulSets(namespace),
+		statefulSetNamespace: namespace,
+	}
+}
+
+func (sc *ScaleCache) GetScale(ctx context.Context, statefulSetName string, options metav1.GetOptions) (*autoscalingv1.Scale, error) {
+	sc.lock.RLock()
+	defer sc.lock.RUnlock()
+
+	if entry, ok := sc.entries.Get(statefulSetName); ok {
+		entry := entry.(scaleEntry)
+		return &autoscalingv1.Scale{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      statefulSetName,
+				Namespace: sc.statefulSetNamespace,
+			},
+			Spec: autoscalingv1.ScaleSpec{
+				Replicas: entry.specReplicas,
+			},
+			Status: autoscalingv1.ScaleStatus{
+				Replicas: entry.statReplicas,
+			},
+		}, nil
+	}
+
+	scale, err := sc.statefulSetClient.GetScale(ctx, statefulSetName, options)
+	if err != nil {
+		return scale, err
+	}
+
+	sc.setScale(statefulSetName, scale)
+
+	return scale, nil
```

Comment:
Let's try to avoid returning the same object as the one in the case, as modifying it would change the value in the cache
```suggestion
	return scale.DeepCopy(), nil
```

Reply:
Since we only store the int values in `.Spec.Replicas` and `.Status.Replicas`, I don't think we are returning the same object that's in the cache